### ISSUE TITLE
HNSW parameters for vector indexing and searching should be configurable

### ISSUE THREAD

**Issue by xionghuaidong** [neo4j/neo4j#13501]
The HNSW vector search algorithm supports two parameters for vector index construction: m and ef_construction. There is no way to set them in Neo4j 5.22.0. I modified the Neo4j source code to make it support the following usage:

```text
CREATE VECTOR INDEX personNameVecIndex IF NOT EXISTS FOR (m:Person) ON m.embedding
OPTIONS {indexConfig: {
  `vector.dimensions`: 2,
  `vector.similarity_function`: 'cosine',
  `vector.hnsw.m`: 18,
  `vector.hnsw.ef_construction`: 240
}};
```

Moreover, ``db.index.vector.queryNodes`` and ``db.index.vector.queryRelationships`` pass ``numberOfNearestNeighbours`` to Lucene directly, which may not result in good accuracy. I modified the Neo4j source code to make it support the following usage:

```text
CALL db.index.vector.queryNodes('personNameVecIndex', 10, [0.866, 0.5], 100)
YIELD node AS person, score
RETURN person.name AS name, score;
```

The modification is available here [xionghuaidong/neo4j, branch ``5.22-hnsw-param``](https://github.com/xionghuaidong/neo4j/tree/5.22-hnsw-param).

**Comment by parnmatt** [comment]
Hi @xionghuaidong thanks for reaching out, and for the efforts you've put into implementing something that hopefully works for your purpose!

---

Indeed you are correct the HNSW hyperparameters are not _currently_ configurable.
The team working on the vector index have wanted such customisation points exposed for a while; however, as is common in all industries, priorities and bandwidth come into play.

You'll be happy to hear that the hyperparameter configuration will be exposed in 5.23 (among quantization and optional/defaulted settings), which should be available to users in a few weeks. I'm glad great minds think alike, you've also chosen the same index-setting _names_ as we did for these two; therefore you should have very little to change in your interface.

Without diving too deeply into your implementation, I believe that any vector index you may have created with your fork could be usable without reconstruction when updating to the implementation we have in the release pipeline. However, please take that with a grain of salt, there's a chance things may not quite line up and work when you upgrade, so be prepared you may need to rebuild any of those indexes.

Our implementations do differ quite a bit as we have to maintain safety across several Enterprise features like rolling upgrades in clustered databases, backwards compatibility both internally and on the user surface, and rendering of the metadata. Such things required versioned validation. From skimming your implementation, I personally quite like its simplicity, it appears it should do the job for your uses in a Community Edition, though perhaps won't render correctly in a few places.
I still would suggest upgrading to our 5.23 implementation once available.

---

In regards to the latter suggestion. As you seem to be familiar with HNSW; `efSearch` needs to be equal to, or larger than `k`. Lucene does the minimum and currently uses `k` as `efSearch`. There was a PR up to add `efSearch` to Lucene that we were hopeful for; however, it was closed recently without merging as they explore other refactors.

Currently you can emulate `efSearch = 100` and `k = 10` with a `LIMIT`.
```cypher
CALL db.index.vector.queryNodes($indexName, 100, $query)
YIELD node, score
LIMIT 10
RETURN node, score;
```
This is intended, and consistent with other patterns in Cypher.

We have had discussions on introducing `efSearch` explicitly in the query procedure, and how to present that in a logical and semantic way, that is backwards compatible. We are also waiting on some movement in ISO standards working groups that are developing SQL and GQL around vectors, as that may influence directions taken on the user surface.

**Comment by xionghuaidong** [comment]
@parnmatt Thanks for your replying. Glad to know the hyperparameter configuration feature will be implemented in 5.23. I will upgrade to 5.23 soon when the feature is available.
